File: intersectionOfArrays.java
// Time Complexity : O(log n) + O(m log n) where n is the length of the smaller array and m is the length of the larger array.
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*
 * 1. Sort both the arrays and perform binary search on larger array.
 * 2. Check if the element in the smaller array is present in the larger array using binary search and also reduce the search space by updating the low pointer.
 * 3. If the element is found, add it to the result list and return the result.
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        
        int n1 = nums1.length, n2 = nums2.length;
        if(n1 > n2)
            intersect(nums2, nums1);
        
        int low = 0, high = n2-1;
        List<Integer> result = new ArrayList<>();

        for(int num : nums1){
            if(firstOccurance(nums2, num, low, high) != -1){
                int newLow = firstOccurance(nums2, num, low, high);
                low = newLow + 1;
                result.add(num);
            }
        }

        int[] res = new int[result.size()];
        for(int i=0; i<result.size(); i++){
            res[i] = result.get(i);
        }

        return res;
    }

    private int firstOccurance(int[] nums, int target, int low, int high){
        while(low <= high){
            int mid = low + (high-low)/2;
            if(nums[mid] == target){
                if(low == mid || nums[mid-1] != target) 
                    return mid;
                high = mid-1;
            }else if(nums[mid] > target){
                high = mid-1;
            }else {
                low = mid+1;
            }
        }
        return -1;
    }
}

File: Intersection of arrays.py
#TC:O(m+n) where m and n are the leng of both arrays
#SC:O(m) size of smaller array
class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        hashMap=dict()
        li=[]
        def intersection(nums2,nums1):
            for i in range(len(nums2)):
                if nums2[i] in hashMap:
                    hashMap[nums2[i]]+=1
                else:
                    hashMap[nums2[i]]=1
            for j in range(len(nums1)):
                if nums1[j] in hashMap:
                    if hashMap[nums1[j]]>0:
                        hashMap[nums1[j]]-=1
                        li.append(nums1[j])
            return li
        if len(nums1)>len(nums2):
            return intersection(nums2,nums1)
        else:
            return intersection(nums1,nums2)

#code using 2 pointer solution
'''
#TC:O(m+n) where m and n are the leng of both arrays
#SC:O(1)
class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        m=0
        n=0
        li=[]
        nums1.sort()
        nums2.sort()
        while(m<len(nums2) and n<len(nums1)):
            if nums2[m]==nums1[n]:
                li.append(nums2[m])
                m+=1
                n+=1
            elif nums2[m]<nums1[n]:
                m+=1
            else:
                n+=1
        return li
'''

#code using binary search
'''
    # TC: O(mlog(n)) if sorted
    # SC: O(1)
class Solution:
    def intersect(self, nums1, nums2):
        if len(nums1)>len(nums2): return self.intersect(nums2,nums1)
        nums1.sort()
        nums2.sort()
        def binarysearch(nums2,l,h,target):
            low=l
            high=h
            while low<=high:
                mid=low+(high-low)//2
                if nums2[mid]==target:
                    if mid==low or nums2[mid-1]!=nums2[mid]:
                        return mid
                    else:
                        high=mid-1
                elif nums2[mid]>target:
                    high=mid-1
                else:
                    low=mid+1
            return -1
        l=0
        h=len(nums2)-1
        result=[]
        for i in nums1:
            idx=binarysearch(nums2,l,h,i)
            #print(idx)
            if idx!=-1:
                result.append(i)
                l=idx+1
        return result
'''
        
                
                
                